Add binary_mclp_distance_matrix into pyspatialopt.models

huanfachen opened the pull request:
A binary MCLP model using a distance matrix

huanfachen committed:
add binary_mclp_distance_matrix in pyspatialopt.models

A binary MCLP model using a distance matrix

apulverizer commented on examples/binary_lscp_arcpy.py:
```diff
@@ -17,18 +18,18 @@
     sh.setFormatter(formatter)
     logger.addHandler(sh)
-
+    scriptDir = os.path.dirname(os.path.realpath(__file__))
```
can you change this to `script_dir` to make it more "pythonic"?

I assume this was to run it from any directory?

huanfachen replied:
Hi. I dont want to commit the changes in the examples/binary_lscp_arcpy.py and binary_mclp_arcpy.py files so I put them in the .gitignore files. However, the .gitignore file does not work here and these changes were submitted. 
Any idea?

apulverizer replied:
Can you make a copy of those files and re-name it something else? Then you can have those file locally for yourself.

Then just copy the original files on top of the ones you modified. Then commit those 2 files. That should make those files unchanged.

Or you can try and revert your last couple commits for those specific files if you're familiar with git.

apulverizer commented on examples/binary_mclp_arcpy.py:
```diff
@@ -17,14 +17,19 @@
     sh.setFormatter(formatter)
     logger.addHandler(sh)
-
+    #import os
```
can you remove these commented out lines?

apulverizer commented on examples/binary_mclp_arcpy.py:
```diff
@@ -17,14 +17,19 @@
+    #import os
+    #print(os.getcwd())
+    import os
+    scriptDir = os.path.dirname(os.path.realpath(__file__))
```
change to `script_dir`

apulverizer commented on examples/binary_mclp_distance_matrix_example.py:
```diff
@@ -0,0 +1,104 @@
+    logger.addHandler(sh)
+
+    import os
+    scriptDir = os.path.dirname(os.path.realpath(__file__))
```
change to `script_dir`

apulverizer reviewed (changes requested):
Thanks for the PR, I haven't actually run it but it looks pretty good. 

Can you clean up the code a bit by removing any commented out code?
There were a couple of other things noted in my comments.

I'm not sure what the Jupyter notebook was trying to show. That would need to be cleaned up a lot if you want that merged, otherwise can you remove that from the PR?

huanfachen committed:
Binary MCLP model using distance matrix

Adding Files to .gitignore (only for testing):
examples/binary_mclp_test_1.ipynb
exmaples/binary_mclp_distance_matrix.py
exmaples/binary_lscp_arcpy.py
exmaples/binary_mclp_arcpy.py

apulverizer commented on .gitignore:
```diff
@@ -69,3 +69,13 @@ target/
 .cpg
+
+#IDE folder
+.vscode
```
keep this line but remove all the others

huanfachen replied:
I donot get it. If so (all the other lines are removed), the IPython notebook file will be committed, which is not my intention.

apulverizer replied:
If I understand correctly, you committed a bunch of files and pushed those changes. When you got to commit the changes you should only add/commit the specific files that you want to add to the repo or update in the repo.

So say I add some new files locally called `test.py` and `test2.py`. And I only want to commit `test.py` then I would just do `git add test.py` and then `git commit -m "add test.py"`  and then git push. Then only `test.py` would be committed and pushed the server. And I don't have to update the `.gitignore` file. 

I'm guessing maybe you committed an entire directory and that included all of the files that were add/changed.

huanfachen replied:
Thanks, now I understand. I figure out how to use git to revert files and remove cache. Please have a look at my latest commit. Let me know if there is anything that needs to be improved.

huanfachen committed:
Add binary_mclp_distance_matrix model. Revert binary_mclp_arcpy and binary_lscp_arcpy

huanfachen committed:
Revert binary_mclp_arcpy and binary_lscp_arcpy

huanfachen committed:
Revert binary_mclp_arcpy and binary_lscp_arcpy

huanfachen committed:
remove models/test/py

huanfachen committed:
Update .gitignore. Only adding one line

apulverizer commented on pyspatialopt/models/binary_mclp_distance_matrix.py:
```diff
@@ -0,0 +1,128 @@
+# -*- coding: UTF-8 -*-
+import logging
+import sys
```
I pulled down your branch and the `sys` and `arcpy` modules are not used, can you remove them?

apulverizer commented on pyspatialopt/models/binary_mclp_distance_matrix.py:
```diff
@@ -0,0 +1,128 @@
+import os
+from pyspatialopt.models import utilities
+from pyspatialopt.models import covering
+
```
Add an extra new line to follow [Pep8](https://www.python.org/dev/peps/pep-0008/) styling guidelines

apulverizer commented on tests/binary_mclp_distance_matrix_test.py:
```diff
@@ -0,0 +1,33 @@
+            res_coverag = binary_mclp_distance_matrix.binary_mclp_distance_matrix(file_distance_matrix, service_dist, num_facility, workspace_path=workspace_path)
+            print("A coverage of {0} is obtained with {1} facilities".format(res_coverag["percent_demand_coverage"], num_facility))
+            self.assertAlmostEquals(res_coverag["percent_demand_coverage"], percent_coverage, places=5)
+
```
Add another new line here for PEP 8

apulverizer commented on pyspatialopt/models/binary_mclp_distance_matrix.py:
```diff
@@ -0,0 +1,128 @@
+        output["totalDemand"] += row["demand"]    
+    logging.getLogger().info("Binary coverage successfully generated.")
+    return output
+
```
Add a new line here for Pep 8

apulverizer commented on pyspatialopt/models/binary_mclp_distance_matrix.py:
```diff
@@ -0,0 +1,128 @@
+    }
+
+    return result_coverage
+    
```
Delete these extra ilnes here. There should be one newline after your `return` statement per PEP 8.

huanfachen replied:
A bit confused here: in pep8 it says "Surround top-level function and class definitions with two blank lines". Looks like two blank lines are needed after the `return` statement?

apulverizer commented on pyspatialopt/models/binary_mclp_distance_matrix.py:
```diff
@@ -0,0 +1,128 @@
+    logging.getLogger().info("Binary coverage successfully generated.")
+    return output
+
+def binary_mclp_distance_matrix(file_distance_matrix, service_dist, num_facility, list_field_req = ["facility_id", "demand_id", "demand", "distance"], facility_variable_name = "facility", workspace_path = "."):
```
I didn't see this before. This is one of those python gotchas. A list is immutable and should never be used as a default value when it's a parameter. You can read more about it [here](https://docs.python-guide.org/writing/gotchas/)

Instead set `list_field_req=None` and then do a check in the function like:
```python
if list_field_req is None:
    list_field_req = ["facility_id", "demand_id", "demand", "distance"]

huanfachen replied:
Good point. Do you mean "a list is mutable and should never be used as a default value"?

apulverizer reviewed (changes requested):
Thanks for making the changes. I took another look and noticed a few style changes that should be made. In addition to the specific comments, in the functions when you set default arguments like:
```python
def some_function(facility_variable_name = "facility", workspace_path = ".")
```
there should not be any spaces around the `=` signs. It should look like:
```python
def some_function(facility_variable_name="facility", workspace_path=".")
```

Otherwise it looks good and I was able to run the test you provided (after I commented out the `import arcpy` line)

If you make the suggested changes I'd be happy to merge it.

huanfachen committed:
Resolve three files. Follow PEP8

apulverizer commented on examples/binary_mclp_distance_matrix_example.py:
```diff
@@ -0,0 +1,94 @@
+    for field in list_field_req:
+        if field not in item_pairwise_distance.keys():
+            logger.info("Error: this field {} not found in the distance csv".format(field))
+            sys.exit(0)
```
sys isn't imported in this file . Sorry, I may have mad you remove that earlier.

apulverizer commented on tests/binary_mclp_distance_matrix_test.py:
```diff
@@ -0,0 +1,32 @@
+            print("A coverage of {0} is obtained with {1} facilities".format(res_coverag["percent_demand_coverage"], num_facility))
+            self.assertAlmostEquals(res_coverag["percent_demand_coverage"], percent_coverage)
+
+    def test_simple_case(self):
```
This function name is duplicated

apulverizer reviewed (changes requested):
Looks good other than those two minor things. Thanks for re-formatting it.

huanfachen committed:
Update binary_mclp_distance_matrix_example.py and binary_mclp_distance_matrix_test.py

apulverizer committed:
fix
